torch_xla/csrc/init_python_bindings.cpp

```diff
@@ -1147,7 +1147,35 @@ void InitXlaModuleBindings(py::module m) {
   m.def("_xla_get_tensor_id",
         [](const at::Tensor& tensor) { return GetTensorId(tensor); });
   m.def("_xla_get_spmd_config_is_locked", []() { return GetLockSpmdConfig(); });
+  m.def("_xla_force_spmd_device", []() {
+    // It is actually more easier to force SPMD mode than blocking if there is
+    // non-SPMD initialized tensors, for the 3rd-party solution integration. For
+    // instance, HuggingFace trainer pre-loads embeddings table and the training
+    // initialization is done over multiple scripts. Being able to force SPMD
+    // allows the users to call `xr.use_spmd()` more freely, given that the
+    // earlier they call, the smaller the one-time overhead of replicating
+    // non-SPMD backed tensors.
+    torch::lazy::BackendDevice backend_device = bridge::GetCurrentDevice();
+    std::vector<XLATensorPtr> xtensors =
+        XLAGraphExecutor::Get()->GetLiveTensors(&backend_device);
+    for (auto xtensor : xtensors) {
+      XlaDeviceType xla_device_type =
+          static_cast<XlaDeviceType>(xtensor->GetDevice().type());
+      if (xla_device_type != XlaDeviceType::SPMD) {
+        // TODO(yeounoh) this can be further optimized via CopyToDevice.
+        at::Tensor tensor = xtensor->ToTensor(false);
```

Comment:
So we are copying the tensor and than shard the copied one? So, there is no way to shard the existing tensors directly?

Reply:
Good question, some things to note and why I decided to do this way
- it's not re-sharding, as we only need to replicate a single PjRtData across devices.
- the data already fits on a single device memory
- it's a one-time overhead when we initialize the SPMD mode, 
- it's cleaner this way, as we drop the reference to the old PjRtData
- as "TODO(yeounoh) this can be further optimized via CopyToDevice" suggests, I implemented `CopyToDevice` API a while back to support device to device transfer. It's an optimization, but given that this is a one-time overhead at the startup time, I deprioritized this.

I actually have a generic re-sharding API for auto-sharding feature, but that runs a separate computation, can be more efficient, but we may have to worry about when stack gets recycled afterward.

Reply:
cc @JackCaoG